### PR TITLE
Extract Henson::APIClient

### PR DESCRIPTION
**Pull request by wfarr**

### CODE REVIEW

File: lib/henson/api_client.rb [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+  class APIClient
+    attr_reader :connection
+
+    def initialize api_host, options = {}
+      @connection = Faraday.new :url => "https://#{api_host}/"
+      @options    = options
+      @tag_cache  = {}
+    end
+
+    def tags_for_repo repository, options = {}
```
Comment (rodjek) [inline lib/henson/api_client.rb]: Given these methods deal directly with the GitHubTarball source, it would seem better to have a generic `Henson::APIClient` class that manages the Faraday connection object (`request`, `handle`) and then a `Henson::Source::GitHubTarball::APIClient` class that inherits it that contains methods specific to the GitHubTarball source (like `tags_for_repo`, `download_tag_for_repo`, etc).

Comment (wfarr) [inline lib/henson/api_client.rb]: For right now, I'm just working on getting this initial extraction working and passing the test suite before continuing on to the other refactoring.

We'll need to eventually have:
- `Henson::API::GitHubClient`
- `Henson::API::PuppetForgeClient`

These classes will then all inherit from `Henson::API::Client`, which will be an internal class mostly gutted from the ongoing work here to gain support for `request` and `handle` and the like.

From there, for persistent connection handle sanity, `Henson.api_client` will inevitably be replaced with `Henson.api_clients`, which will be a cache layer to make sure we only keep a single API client per type and per host. This will allow us to handle multiple sources of the same backend type elegantly and efficiently.

**Comment by wfarr** [comment]
Items affected by this that need test coverage restored:

screen shot 2013-04-30 at 2 41 22 pm

**Comment by wfarr** [comment]
Okay, `Henson::API::GitHubClient` needs a bit of love, and we can wrap some more tests around `Henson::API::Client#request` as we need to.

**Comment by wfarr** [comment]
Ready for :eyes:

File: spec/henson/api/client_spec.rb [hunk 2]
```diff
@@ -0,0 +1,92 @@
+require "spec_helper"
+
+require "henson/api/client"
```
Comment (rodjek) [inline spec/henson/api/client_spec.rb:3]: Unnecessary, `spec_helper.rb` requires `henson`

File: spec/henson/api/client_spec.rb [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+require "json"
```
Comment (rodjek) [inline spec/henson/api/client_spec.rb:5]: Unnecessary, `lib/henson/api/client.rb` requires `multi_json`

File: spec/henson/api/github_client_spec.rb [hunk 4]
```diff
@@ -0,0 +1,61 @@
+require "spec_helper"
+
+require "henson/api/github_client"
```
Comment (rodjek) [inline spec/henson/api/github_client_spec.rb:3]: Unnecessary

File: lib/henson/api/client.rb [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+      #
+      # Returns a Hash.
+      def handle response, request_options = {}
+
+        if response.success?
+
+          if response.body.empty?
+            { "ok" => true }
+          else
+            case response.env[:response_headers]["content-type"]
```
Comment (rodjek) [inline lib/henson/api/client.rb:60]: This could be condensed to a simple `if` statement given it only has a single case and default.

``` ruby
if response.env[:response_headers]['content-type'] =~ /gzip/
  response.body
else
  MultiJson.load(response.body)
end
```

File: lib/henson/api/client.rb [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+
+      # Internal: Handle an HTTP response body as JSON.
+      #
+      # response - The Faraday::Response generated from the Faraday::Request.
+      #
+      # Returns a Hash.
+      def handle response, request_options = {}
+
+        if response.success?
+
```
Comment (rodjek) [inline lib/henson/api/client.rb:56]: Unnecessary whitespace?

File: spec/henson/api/client_spec.rb [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+      expect(lambda { described_class.new }).to \
+        raise_error(ArgumentError)
+
+      expect(described_class.new "foo.test").to \
+        be_a(described_class)
+    end
+  end
+
+  describe "#request" do
+  end
```
Comment (rodjek) [inline spec/henson/api/client_spec.rb:21]: Tests for `#request` forgotten?

**Comment by wfarr** [comment]
All these fixes are addressed in a follow-up branch. Closing this PR before it like triples in size.
